[Flang][Driver] Add a flag to control zero initialization of global v…

kiranchandramohan opened the pull request:
…ariables

Patch adds a flag to control zero initialization of global variables without default initialization. The default is to zero initialize.

kiranchandramohan committed:
[Flang][Driver] Add a flag to control zero initialization of global variables

Patch adds a flag to control zero initialization of global variables
without default initialization. The default is to zero initialize.

tarunprabhu reviewed (approved):
LGTM. Thanks Kiran.

kiranchandramohan commented:
> LGTM. Thanks Kiran.

Is the name of the flag alrite?

tarunprabhu commented:
Perhaps leave this up for a bit longer to see if we can think of something better. Or others may have some suggestions

jeanPerier reviewed (approved):
Thanks Kiran!
Regarding the name, I never have strong opinions about them. Maybe `fno-zero-init-globals` is enough (the description talks about the without-init part, and it is rather obvious I think that this will not affect/control user define initialization). I also like `-fno-init-uninitialized-globals`.

gfortran has `-finit-local-zero` which flang will likely eventually have. It has no such options for globals (it is putting everything is .bss ASFAIK). Maybe `-fno-init-global-zero` would be consistent whit that existing name. 

Maybe ask in the flang slack channel for opinions?

carlocab commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3494,6 +3494,9 @@ def fno_struct_path_tbaa : Flag<["-"], "fno-struct-path-tbaa">, Group<f_Group>;
 def fno_strict_enums : Flag<["-"], "fno-strict-enums">, Group<f_Group>;
 def fno_strict_overflow : Flag<["-"], "fno-strict-overflow">, Group<f_Group>,
   Visibility<[ClangOption, FlangOption]>;
+def fno_zero_init_global_without_init : Flag<["-"], "fno-zero-init-global-without-init">, Group<f_Group>,
```
Following https://github.com/llvm/llvm-project/pull/121619#discussion_r1906391568, I suggest considering `-fskip-zero-init-globals` or similar.

vitalybuka replied:
This looks like regular boolean flag which should support `clang -fflag -fno-flag -fflag -fno-flag -fflag -fno-flag` invocation. Id rather add positive version for consistency.

There have non-boolean settings https://github.com/llvm/llvm-project/pull/121619#discussion_r1906391568 
we just have no idea what could be a meaning difference for `-fno-sanitize-top-hot=null=0.5` vs `-fsanitize-top-hot=null=0.5`.

kiranchandramohan replied:
Added a positive version as well.

kiranchandramohan committed:
Change name to init-global-zero and add positive flag as well

kiranchandramohan commented:
> -fno-init-global-zero

Thanks @tarunprabhu @jeanPerier, Changed to `fno-init-global-zero` and also added a positive version`-fno-init-global-zero`.

jeanPerier reviewed (approved):
Beware there is a formatting issue (see github CI). LGTM otherwise, thanks Kiran!

carlocab commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3494,6 +3494,12 @@ def fno_struct_path_tbaa : Flag<["-"], "fno-struct-path-tbaa">, Group<f_Group>;
 def fno_strict_enums : Flag<["-"], "fno-strict-enums">, Group<f_Group>;
 def fno_strict_overflow : Flag<["-"], "fno-strict-overflow">, Group<f_Group>,
   Visibility<[ClangOption, FlangOption]>;
+defm init_global_zero : BoolFOption<"init-global-zero",
```
`-fzero-init-globals`, maybe? Feel free to ignore if you disagree, though.

kiranchandramohan replied:
Thanks @carlocab. As Jean pointed out, there is a gfortran flag `-finit-local-zero`. So the name `-finit-global-zero` was chosen to be similar to it.
Hence, I am not taking this comment. Hope that is OK.

kiranchandramohan committed:
Fix formatting

kiranchandramohan committed:
Switch to BoolOptionWithoutMarshalling

Flang does not use the Marshalling fields and the setting of
CodegenOpts from Options.td.

kiranchandramohan committed:
Merge branch 'main' into zero-init

kiranchandramohan committed:
Fix formatting

kiranchandramohan committed:
Merge branch 'main' into zero-init

kiranchandramohan committed:
Merge branch 'main' into zero-init

kiranchandramohan committed:
Minor syntax changes

kiranchandramohan committed:
Correct help description
